src/ssl.c

```diff
@@ -19676,6 +19738,84 @@ int wolfSSL_X509_cmp_current_time(const WOLFSSL_ASN1_TIME* asnTime)
 }
 #endif
 
+/* return -1 if asnTime is earlier than or equal to cmpTime, and 1 otherwise
+ * return 0 on error
+ */
+int wolfSSL_X509_cmp_time(const WOLFSSL_ASN1_TIME* asnTime, time_t* cmpTime)
+{
+    int ret = 0, i = 0;
+    time_t tmpTime, *pTime = &tmpTime, aTime;
+    byte data_ptr[MAX_TIME_STRING_SZ], inv = 0;
+    struct tm ts;
+
+    if (asnTime != NULL) {
+        if (cmpTime == NULL) {
+            /* Use current time */
+            *pTime = XTIME(0);
+        }
+        else {
+            pTime = cmpTime;
+        }
+
+        /* Convert ASN1_time to time_t */
+        XMEMSET(&ts, 0, sizeof(struct tm));
+
+        /* Check type */
+        if (asnTime->data[0] == ASN_UTC_TIME) {
+            /* 2-digit year */
+            i += 2; /* Skip type and size */
+            XMEMCPY(data_ptr, &asnTime->data[i], ASN_UTC_TIME_SIZE);
+            ts.tm_year = (data_ptr[i] - '0') * 10; i++;
+            ts.tm_year += data_ptr[i] - '0'; i++;
+            if (ts.tm_year < 70) {
+                ts.tm_year += 100;
+            }
+        }
+        else if (asnTime->data[0] == ASN_GENERALIZED_TIME) {
+            /* 4-digit year */
+            i += 2; /* Skip type and size */
+            XMEMCPY(data_ptr, &asnTime->data[i], ASN_GENERALIZED_TIME_SIZE);
+            ts.tm_year = (data_ptr[i] - '0') * 1000; i++;
+            ts.tm_year += (data_ptr[i] - '0') * 100; i++;
+            ts.tm_year += (data_ptr[i] - '0') * 10; i++;
+            ts.tm_year += data_ptr[i] - '0'; i++;
+            ts.tm_year -= 1900;
+        }
+        else {
+            /* Invalid type */
+            inv = 1;
+        }
+
+        if (inv != 1) {
+            ts.tm_mon = (data_ptr[i] - '0') * 10; i++;
+            ts.tm_mon += (data_ptr[i] - '0') - 1; i++; /* January is 0 not 1 */
+            ts.tm_mday = (data_ptr[i] - '0') * 10; i++;
+            ts.tm_mday += (data_ptr[i] - '0'); i++;
+            ts.tm_hour = (data_ptr[i] - '0') * 10; i++;
+            ts.tm_hour += (data_ptr[i] - '0'); i++;
+            ts.tm_min = (data_ptr[i] - '0') * 10; i++;
+            ts.tm_min += (data_ptr[i] - '0'); i++;
+            ts.tm_sec = (data_ptr[i] - '0') * 10; i++;
+            ts.tm_sec += (data_ptr[i] - '0');
+
+            /* Convert to Epoch time */
```

Comment:
We have the static function DateGreaterThan in wolfcrypt/src/asn.c that could be made a WOLFSSL_LOCAL function and used here.

XGMTIME would convert the time_t to a struct tm
And ExtractDate in wolfcrypt/src/asn.c can convert the date string to a struct tm. See wolfSSL_ASN1_TIME_to_string -> GetTimeString -> ExtractDate

Reply:
Yes, that is a much easier method.